crates/sui-package-dump/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    collections::BTreeMap,
+    fs,
+    path::{Path, PathBuf},
+};
+
+use anyhow::{bail, ensure, Context, Result};
+use client::Client;
+use fastcrypto::encoding::{Base64, Encoding};
+use query::{limits, packages};
+use sui_types::object::Object;
+use tracing::info;
+
+mod client;
+mod query;
+
+/// Ensure all packages created before `before_checkpoint` are written to the `output_dir`ectory,
+/// from the GraphQL service at `rpc_url`.
+///
+/// `output_dir` can be a path to a non-existent directory, in which case it will be created, or an
+/// existing empty directory (in which case it will be filled), or an existing directory that has
+/// been written to in the past (in which case this invocation will pick back up from where the
+/// previous invocation left off).
+pub async fn dump(
```

Comment:
For my understanding, in other words, dump all the packages that have been created by `before_checkpoint`